### ISSUE TITLE
Error during build: randomizer

### ISSUE THREAD

**Issue by LDamer** [ofloveandhate/bertini_real#86]
I followed the instructions on how to install bertini_real, with bertini already installed.
I cloned the git repo, used `libtoolize`, `autoreconf -i,` `./configure` and it all worked fine.
But when I try to build with `make -j 4`, I get errors regarding a _randomizer_ in the files at `./include/decompositions`. The entire output to stderr is in the log file [make_error.log](https://github.com/user-attachments/files/17019525/make_error.log).

Hope someone can help:)

**Comment by LDamer** [comment]
System: Ubuntu 24.04.1 LTS (freshly installed)

**Comment by LDamer** [comment]
`bertini input` itself works fine with the cayley cubic example from the manual.

**Comment by ofloveandhate** [comment]
Thanks for raising an issue!!!  As a step to help me get in sync with you, will you describe what "freshy installed" means?  Did you install packages using apt?  Which ones?

**Comment by ofloveandhate** [comment]
To debug this, I will spin up a 24.04.1 virtual machine on my mac.  I haven't built under 24.04 yet.

**Comment by LDamer** [comment]
Thanks for your answer!
I  re-installed Ubuntu three weeks ago and just installed some standard stuff. So with "freshly installed" i meant there shouldn't be any weird configs/paths/dependencies/... set. At least I hope so. Nonetheless, you can find the entire apt installation history here :[apt_installs.txt](https://github.com/user-attachments/files/17079444/apt_installs.txt)

As you can see, I started installing bison and flex on 16th September as required for bertini.

**Comment by LDamer** [comment]
**Update**: Here are instructions to reproduce the issue aka how I tried to install it on a fresh Ubuntu.


```
1) install Ubuntu 24.04.1 LTS (default selection, third party software, additional media formats)

2) use apt to install build-essential, libgmp-dev, libmpfr-dev, libopenmpi-bin, flex, bison, git

3) download bertini source code (https://bertini.nd.edu/download.html) and extract archive
	./configure with CXXFLAGS and LDFLAG as in mpicc -showme
	make -j 4
	sudo make install

4) try to install bertini_real
	git clone https://github.com/ofloveandhate/bertini_real
	cd bertini_real
	libtoolize
	autoreconf -i
	./configure with CXXFLAGS and LDFLAG as in mpicc -showme
	make -j 4 	#THIS YIELDS THE ERROR:/ 
```

**Comment by ofloveandhate** [comment]
Thank you very much.  I'm going to try to reproduce your issue.  I just started a new job, so please be patient as most of my time is in some major setup tasks that will conclude soon.  

I'm going to use the Desktop Ubuntu, not the server one.  If that's wrong, please let me know.

**Comment by ofloveandhate** [comment]
I installed Ubuntu 24.04 for ARM in Parallels on my mac.  I installed the libraries and tools via apt as you described.  I couldn't immediately get ./configure to work, so I ran `autoreconf -vfi` in the folder.  Configure worked, and I didn't need to set anything for CPPFLAGS or LDFLAGS, and it's building the parallel version of Bertini 1.  Bertini 1 installed just fine.

Trying to install bertini_real.  `./configure` didn't succeed without passing options for finding mpi.  I used

```
./configure CPPFLAGS=-I/usr/lib/aarch64-linux-gnu/openmpi/include LDFLAGS="-L/usr/lib/aarch64-linux-gnu/openmpi/lib -lmpi"
```

Configuring still failed, complaining about missing Boost.  `sudo apt install libboost-dev` installed Boost 1.83.  Getting bertini_real to find boost correctly is a strugglefest for me right now.  Their m4 macros are not super reliable.  I need to just switch to using CMake, but I can't do that at just this minute. 

Ah, the problem is that i installed libboost-dev not libboost-all-dev.  Fine.  The "library not found" messages from their M4 macros leave a lot to be desired.  `sudo apt install libboost-all-dev` got me through configure with the above block-commented configure line.

---

now to the build step, which is what you asked for help with.  Indeed, I get an error.  

```
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I./include -D_HAVE_MPI -I/usr/include -I/usr/lib/aarch64-linux-gnu/openmpi/include -g -O2 -MT src/decompositions/decomposition.lo -MD -MP -MF src/decompositions/.deps/decomposition.Tpo -c src/decompositions/decomposition.cpp  -fPIC -DPIC -o src/decompositions/.libs/decomposition.o
In file included from src/decompositions/decomposition.cpp:1:
./include/decompositions/decomposition.hpp:267:14: error: 'shared_ptr' in namespace 'std' does not name a template type
  267 |         std::shared_ptr<SystemRandomizer> randomizer() const
      |              ^~~~~~~~~~
./include/decompositions/decomposition.hpp:7:1: note: 'std::shared_ptr' is defined in header '<memory>'; did you forget to '#include <memory>'?
    6 | #include "containers/vertex_set.hpp"
  +++ |+#include <memory>
    7 | 
./include/decompositions/decomposition.hpp:602:14: error: 'shared_ptr' in namespace 'std' does not name a template type
  602 |         std::shared_ptr<SystemRandomizer> randomizer_; ///< the randomizer for the Decomposition.
      |              ^~~~~~~~~~
./include/decompositions/decomposition.hpp:602:9: note: 'std::shared_ptr' is defined in header '<memory>'; did you forget to '#include <memory>'?
  602 |         std::shared_ptr<SystemRandomizer> randomizer_; ///< the randomizer for the Decomposition.
      |         ^~~
./include/decompositions/decomposition.hpp: In member function 'void Decomposition::clone(const Decomposition&)':
./include/decompositions/decomposition.hpp:509:23: error: 'class Decomposition' has no member named 'randomizer_'
  509 |                 this->randomizer_ = std::make_shared<SystemRandomizer>( *other.randomizer_ );
      |                       ^~~~~~~~~~~
./include/decompositions/decomposition.hpp:509:42: error: 'make_shared' is not a member of 'std'
  509 |                 this->randomizer_ = std::make_shared<SystemRandomizer>( *other.randomizer_ );
      |                                          ^~~~~~~~~~~
./include/decompositions/decomposition.hpp:509:42: note: 'std::make_shared' is defined in header '<memory>'; did you forget to '#include <memory>'?
./include/decompositions/decomposition.hpp:509:70: error: expected primary-expression before '>' token
  509 |                 this->randomizer_ = std::make_shared<SystemRandomizer>( *other.randomizer_ );
      |                                                                      ^
./include/decompositions/decomposition.hpp:509:80: error: 'const class Decomposition' has no member named 'randomizer_'
  509 |                 this->randomizer_ = std::make_shared<SystemRandomizer>( *other.randomizer_ );
      |                                                                                ^~~~~~~~~~~
./include/decompositions/decomposition.hpp: In member function 'void Decomposition::init()':
./include/decompositions/decomposition.hpp:613:17: error: 'randomizer_' was not declared in this scope; did you mean 'randomize_SLP'?
  613 |                 randomizer_ = std::make_shared<SystemRandomizer> ();
      |                 ^~~~~~~~~~~
      |                 randomize_SLP
./include/decompositions/decomposition.hpp:613:36: error: 'make_shared' is not a member of 'std'
  613 |                 randomizer_ = std::make_shared<SystemRandomizer> ();
      |                                    ^~~~~~~~~~~
./include/decompositions/decomposition.hpp:613:36: note: 'std::make_shared' is defined in header '<memory>'; did you forget to '#include <memory>'?
./include/decompositions/decomposition.hpp:613:64: error: expected primary-expression before '>' token
  613 |                 randomizer_ = std::make_shared<SystemRandomizer> ();
      |                                                                ^
./include/decompositions/decomposition.hpp:613:67: error: expected primary-expression before ')' token
  613 |                 randomizer_ = std::make_shared<SystemRandomizer> ();
      |                                                                   ^
./include/decompositions/decomposition.hpp: In member function 'void Decomposition::copy(const Decomposition&)':
./include/decompositions/decomposition.hpp:644:23: error: 'class Decomposition' has no member named 'randomizer_'
  644 |                 this->randomizer_ = other.randomizer_;
      |                       ^~~~~~~~~~~
./include/decompositions/decomposition.hpp:644:43: error: 'const class Decomposition' has no member named 'randomizer_'
  644 |                 this->randomizer_ = other.randomizer_;
      |                                           ^~~~~~~~~~~
src/decompositions/decomposition.cpp: In member function 'void Decomposition::send(int, ParallelismConfig&) const':
src/decompositions/decomposition.cpp:545:9: error: 'randomizer_' was not declared in this scope; did you mean 'randomize_SLP'?
  545 |         randomizer_->send(target,mpi_config);
      |         ^~~~~~~~~~~
      |         randomize_SLP
src/decompositions/decomposition.cpp: In member function 'void Decomposition::receive(int, ParallelismConfig&)':
src/decompositions/decomposition.cpp:633:9: error: 'randomizer_' was not declared in this scope; did you mean 'randomize_SLP'?
  633 |         randomizer_->receive(source,mpi_config);
      |         ^~~~~~~~~~~
      |         randomize_SLP

```

The cause appears to be that I don't have `<memory>` included.  I'll try including it.  Indeed, including memory allowed compilation to succeed.  Now I just need to commit and push.

I need to adjust the configure line, the `-lmpi` wasn't in the right place, and since we did `mpicc --showme`, not `mpicxx --showme`, we were missing the c++ mpi library.  This is the one that succeeded:

```
./configure CPPFLAGS="-I/usr/lib/aarch64-linux-gnu/openmpi/include -I/usr/lib/aarch64-linux-gnu/openmpi/include/openmpi" LDFLAGS="-L/usr/lib/aarch64-linux-gnu/openmpi/lib"  LIBS="-lmpi -lmpi_cxx"
```

Aside: I'm frustrated that Bertini1 found MPI without my assistance, but Bertini_real didn't.  That's not this issue, so I won't discuss it further here.

I installed bertini_real.  

I'm trying to run `bertini`, but after launching it stalls, likely due to some MPI configuration thing.  I don't know. But this seems like an unrelated issue, so I won't discuss it further here.

---

Bottom line: adding `#include <memory>` to `decomposition.hpp` allowed compilation to succeed.

**Comment by ofloveandhate** [comment]
Take a pull and try to compile.  I bet you're good to go.

**Comment by LDamer** [comment]
**Issue resolved!**
Thanks for your help! It build just fine:)

**Comment by ofloveandhate** [comment]
I hope you enjoy using the software.  Thanks for working through this with me!
